plugins/catalog-backend/src/ingestion/descriptors/types.ts

```diff
@@ -15,18 +15,93 @@
  */
 
 import { ParserOutput } from '../types';
-import { DescriptorEnvelope } from './DescriptorEnvelopeParser';
 
-export type KindParser = {
+/**
+ * The format envelope that's common to all versions/kinds.
+ *
+ * @see https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/
+ */
+export type DescriptorEnvelope = {
+  /**
+   * The version of specification format for this particular entity that
+   * this is written against.
+   */
+  apiVersion: string;
+
   /**
-   * Parses and validates a single envelope into its materialized kind.
+   * The high level entity type being described.
+   */
+  kind: string;
+
+  /**
+   * Optional metadata related to the entity.
    *
-   * These parsers may assume that the envelope is already validated and
-   * well formed.
+   * @see https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#objectmeta-v1-meta
+   */
+  metadata?: {
+    /**
+     * A globally unique ID for the entity. This field can not be set by the
+     * user at creation time, and the server will reject an attempt to do so.
+     * The field will be populated in read operations. The field can
+     * (optionally) be specified when performing update or delete operations,
+     * but the server is free to reject requests that do so in such a way that
+     * it breaks semantics.
+     */
+    uid?: string;
+
+    /**
+     * A positive nonzero number that indicates the current generation of data
+     * for this entity; the value is incremented for each update operation.
+     * This field can not be set by the user at creation time, and the server
+     * will reject an attempt to do so. The field will be populated in read
+     * operations. The field can (optionally) be specified when performing
+     * update or delete operations, and the server will then reject the
+     * operation if it does not match the current stored value.
+     */
+    generation?: number;
```

Comment:
Would like to add to the description that it diverges from the k8s generation in that it will also be incremented by changes to the metadata as well, because I'm assuming we'll want that?

This also made me realize the difference between spec and metadata. Referring back to raghu's article, the spec is the core model, and if we're following that, the catalog service should only consume data from the spec.

Reply:
Not sure about the generation. I was frankly not aware of the distinction. That's interesting, hmmm I have to get back to you on that one. We probably do not want to diverge from the spec but at the same time we MAY want external atomicity for metadata updates too.

How do you mean with consume? The service itself doesn't use any of the fields particularly, right?

Reply:
Current thinking: introducing an etag instead that does not break the generation field expectations

Reply:
Introduced etag, 👀

Reply:
@freben I meant that the catalog service should not be looking at the labels or annotations, just forwarding them. Now imma have a look at etags :grin: